compiler/noirc_driver/src/abi_gen.rs

```diff
@@ -20,15 +23,104 @@ pub(super) fn gen_abi(
     error_types: BTreeMap<ErrorSelector, Type>,
 ) -> Abi {
     let (parameters, return_type) = compute_function_abi(context, func_id);
-    let return_type = return_type
-        .map(|typ| AbiReturnType { abi_type: typ, visibility: return_visibility.into() });
+    let return_type = return_type.map(|typ| AbiReturnType {
+        abi_type: typ,
+        visibility: to_abi_visibility(return_visibility),
+    });
     let error_types = error_types
         .into_iter()
-        .map(|(selector, typ)| (selector, AbiErrorType::from_type(context, &typ)))
+        .map(|(selector, typ)| (selector, build_abi_error_type(context, &typ)))
         .collect();
     Abi { parameters, return_type, error_types }
 }
 
+fn build_abi_error_type(context: &Context, typ: &Type) -> AbiErrorType {
+    match typ {
+        Type::FmtString(len, item_types) => {
+            let length = len.evaluate_to_u32().expect("Cannot evaluate fmt length");
+            let Type::Tuple(item_types) = item_types.as_ref() else {
+                unreachable!("FmtString items must be a tuple")
+            };
+            let item_types =
+                item_types.iter().map(|typ| abi_type_from_hir_type(context, typ)).collect();
+            AbiErrorType::FmtString { length, item_types }
+        }
+        _ => AbiErrorType::Custom(abi_type_from_hir_type(context, typ)),
+    }
+}
+
+pub(super) fn abi_type_from_hir_type(context: &Context, typ: &Type) -> AbiType {
+    // Note; use strict_eq instead of partial_eq when comparing field types
+    // in this method, you most likely want to distinguish between public and private
```

Comment:
```suggestion
```
This comment seems outdated since we don't have `pub`/`priv` information in the types anymore